### DISCUSSION TITLE
Support for smoother gradients

### DISCUSSION THREAD

**Discussion by sona-tau** [YaLTeR/niri#348] [Ideas]
Hello all!

Today I was trying out using different border gradients and a thing I noticed was that the gradients seem to skip some colors. This makes sense (seeing the comment in the default config supplied) since the gradients are based off CSS's `linear-gradient`. This means that the gradient smoothly interpolate colors that go from the `from` color to the `to` color. This makes it so that if you make a gradient between red and blue, the colors go from red, to grayish purple, to blue. The hues in between blue and red get skipped.

The way CSS gets around this is by letting you give `linear-gradient` more arguments. Those arguments being the extra colors that you want the gradient to pass through. As a consequence you can make it go through the missing hues you needed.

My question now is:
- Could there be support for something like what CSS does with the extra arguments to `linear-gradient`?
- Could there be support for a version of `active`/`inactive-gradient` that interpolates the hue of the color instead of the RGB values?

Thank you for your time reading this.

Here are some examples of what I mean:
`linear-gradient(red, blue)`
![image](https://github.com/YaLTeR/niri/assets/65361987/018ec49a-5802-49cf-88ce-862312562672)

`linear-gradient(red, orange, yellow, green, blue)`
![image](https://github.com/YaLTeR/niri/assets/65361987/44683b5d-edee-46a8-99d8-9f0727969b5d)

**Comment by YaLTeR** [comment]
I was thinking to leave the config syntax as is, but add the ability to use a custom border "background image" so to say (so you save  the gradient or anything else in an image and it uses that), or maybe just a custom border shader. Because I don't imagine this is a very widespread need, and these options will allow maximum customizability for the people who want it.

**Reply by mrwensveen** [reply]
Hi. I was going to request the same. The CSS linear-gradient function lets you supply as many stops as you want, but Niri only allows "from" and "to". To maintain backward compatibility a new property "stops" could be introduced that can be used instead of, or in addition to, the existing properties. An array or a comma-delimited string would work.

Psueudo code: `real_stops = [from, ...stops, to].filter(v => v != null)`

If you're interested, I could look at creating a PR.
